grails-app/migrations/core/2020-09-18-convert-API-keys-to-Ethereum-IDs.groovy

```diff
@@ -0,0 +1,122 @@
+		userId,
+		IntegrationKey.Service.ETHEREUM_ID.name(),
+		accountAddress,
+		now,
+		now
+	]);
+}
+
+def createInboxStream(long userId, String accountAddress, Sql sql) {
+	Date now = new Date();
+	sql.executeInsert('INSERT INTO stream (version, id, name, date_created, last_updated, partitions, require_signed_data, auto_configure, storage_days, inbox, inactivity_threshold_hours, example_type, require_encrypted_data) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)', [
+		0,
+		accountAddress,
+		accountAddress,
+		now,
+		now,
+		1,
+		false,
+		false,
+		Stream.DEFAULT_STORAGE_DAYS,
+		true,
+		Stream.DEFAULT_INACTIVITY_THRESHOLD_HOURS,
+		ExampleType.NOT_SET.value,
+		false
+	]);
+	List<Permission.Operation> permissionOperations = [
+		Permission.Operation.STREAM_GET,
+		Permission.Operation.STREAM_EDIT,
+		Permission.Operation.STREAM_DELETE,
+		Permission.Operation.STREAM_PUBLISH,
+		Permission.Operation.STREAM_SUBSCRIBE,
+		Permission.Operation.STREAM_SHARE,
+	];
+	permissionOperations.each { op ->
+		sql.executeInsert('INSERT INTO permission (version, operation, stream_id, anonymous, user_id) VALUES (?, ?, ?, ?, ?)', [
+			0,
+			op.id,
+			accountAddress,
+			false,
+			userId
+		]);
+	}
+}
+
+databaseChangeLog = {
+
+	// For each API key that is attached to a user:
+	// - create a new IntegrationKey for the user
+	// - delete the key
+	changeSet(author: "teogeb", id: "convert-API-keys-to-Ethereum-IDs-1") {
+		grailsChange {
+			change {
+				sql.eachRow('SELECT id, name, user_id FROM `key` where user_id is not null') { row ->
+					String apiKeyId = row['id']
+					String apiKeyName = row['name']
+					int userId = row['user_id']
+					String accountAddress = getAccountAddress(apiKeyId)
+					String integrationKeyName = 'Converted from API key: ' + apiKeyName
+					createIntegrationKey(integrationKeyName, accountAddress, userId, sql)
+				}
+			}
+		}
+		sql('DELETE FROM `key` where user_id is not null')
+	}
+
+	// For each anonymous API key:
+	// - create a new Ethereum user (and inbox stream for that user)
+	// - migrate the key’s permissions to that user
+	// - delete the key
+	changeSet(author: "teogeb", id: "convert-API-keys-to-Ethereum-IDs-2") {
+		grailsChange {
+			change {
+				sql.eachRow('SELECT id, name, user_id FROM `key` where user_id is null') { row ->
+					String apiKeyId = row['id']
+					String accountAddress = getAccountAddress(apiKeyId)
+					def insertResult = sql.executeInsert("INSERT INTO user (version, account_expired, account_locked, enabled, name, password, password_expired, username, signup_method) VALUE (?, ?, ?, ?, ?, ?, ?, ?, ?)", [
+						0,
+						false,
+						false,
+						true,
+						"Anonymous User",
+						AlphanumericStringGenerator.getRandomAlphanumericString(32),
+						false,
+						accountAddress,
+						SignupMethod.MIGRATED.name()
+					]);
+					int userId = insertResult[0][0]
+					sql.execute("UPDATE permission SET user_id = ?, key_id = null WHERE key_id = ?", [ userId, apiKeyId])
```

Comment:
Here or in a separate migration that happens before this one, you might want to fix the issue with anonymous keys having `stream_edit` and `stream_delete` permissions

Reply:
Ok, we'll do BACK-61 before this one.